File: highway_constr.cpp
#include <bits/stdc++.h>
using namespace std;
#define MOD 1000000009

long long int modPower(long long int a , long long int b)
{
    if(b == 0)
        return 1;
    long long int result = modPower(a,b/2);
    result = (result * result) % MOD;

    if(b%2 != 0)
        result = (result * a)%MOD;
    return result;
}

//This is Correct
void countP(int n, int k,long long int dp[][1001])
{
  for (int i=0; i<=n; i++)
    for(int j = 0; j <= k ; j++)
        dp[i][j] = 0;

  for (int i=1; i<=n; i++)
     for (int j=1; j<=i; j++)
       if (j == 1 || i == j)
          dp[i][j] = 1;
       else
          dp[i][j] = ((j*dp[i-1][j])%MOD + (dp[i-1][j-1])%MOD)%MOD;
}

//This is correct
long long int fallingFact(long long int n,int k)
{
    if(k == 0)
        return 1;
    if( k == 1)
        return n%MOD;
    long long int ans = n;
    long long int res = 1;
    long long int next;
    for(int i = 1; i < k ; i++)
    {
        res = ans%MOD;
        next = (n-i)%MOD;
        ans = (res * next)%MOD;
        // if(k<5)
        // cout<<"FAll Fact: "<<k<<" "<<res<<" "<<next<<" "<<ans<<endl;
    }
    return ans;
}

long long int highwayConstruction_Sterling(long long int n, int k) {
    // Complete this
    long long int sum = (n+1)%MOD;
    long long int stirling_mat[1001][1001];
    countP(1000,1000,stirling_mat);
    long long int res = 0;
    for(int i = 0; i <= k ; i++)
    {
        // cout<<k<<" "<<i<<" "<<stirling_mat[k][i]<<endl;
        long long int ff = fallingFact(n,i);
        long long int sm = stirling_mat[k][i];
        // long long smallProd = (ff * sm)%MOD;
        // long long smallDiv = (smallProd*modPower(i+1,MOD-2))%MOD;
        long long int modInverse = modPower(i+1,MOD-2)%MOD;
        long long int smallDiv = ((ff%MOD) * (modInverse%MOD))%MOD;
        long long int smallProd = ((sm%MOD) * (smallDiv%MOD))%MOD;
        res = (res%MOD + smallProd%MOD)%MOD;
        // if(i < 5)
        //     cout<<ff<<" "<<modInverse<<" "<<sm<<" "<<smallDiv<<" "<<smallProd<<" "<<res<<endl;
    }

    sum = ((sum%MOD) * (res%MOD))%MOD;
    sum = (sum%MOD - 1 - modPower(n,k) + MOD)%MOD;
    return sum;
}

int main() 
{
    int q;
    cin >> q;
    for(int a0 = 0; a0 < q; a0++){
        long long int n;
        int k;
        cin >> n >> k;
        if(n <= 1)
        {
            cout<<0<<endl;
            continue;
        }
        long long int result = highwayConstruction_Sterling(n, k);
        cout << result << endl;
    }
    return 0;
}

File: Week-Of-Code-35/highway_constr.py
#!/bin/python

import sys
MOD = 1000000009

stirlingMat = [[0 for x in range(1001)] for y in range(1001)] 

def modPower(a,b):
    if(b == 0):
        return 1
    result = modPower(a,b/2)
    result = (result*result)%MOD
    if(b%2 != 0):
        result = (result * a)%MOD
    return result 

def fallingFactorial(n,k):

    if(k == 0):
        return 1
    if(k == 1):
        return n
    ans = n
    for i in range(1,k):
        ans = ((ans%MOD)*((n-i)%MOD))%MOD
    return ans

def stirlingFunction(n,k):
    for i in range(n+1):
        stirlingMat[i][0] = 0
    for i in range(k+1):
        stirlingMat[0][i] = 0
    for i in range(1,n+1):
        for j in range(1,i+1):
            if( j == 1 or i == j):
                stirlingMat[i][j] = 1
            else:
                 stirlingMat[i][j] = ((j*stirlingMat[i-1][j])%MOD + (stirlingMat[i-1][j-1])%MOD)%MOD

def highwayConstruction(n, k):
    # Complete this function
    if( n <= 1):
        return 0
    SUM = n+1
    res = 0
    for i in range(1,k+1):
        ff = fallingFactorial(n,i)
        sm = stirlingMat[k][i]
        smallDiv = (ff * modPower(i+1,MOD-2))%MOD
        smallProd = (smallDiv * sm)%MOD
        res = ((res%MOD) + (smallProd%MOD))%MOD
    
    SUM = (SUM * res)%MOD
    SUM = (SUM - 1 - modPower(n,k) + MOD)%MOD
    return SUM%MOD

if __name__ == "__main__":
    q = int(raw_input().strip())
    stirlingFunction(1000,1000)
    for a0 in xrange(q):
        n, k = raw_input().strip().split(' ')
        n, k = [long(n), int(k)]
        # print fallingFactorial(10,5)
        # for i in range(10):
        #     for j in range(10):
        #         print stirlingMat[i][j],
        #     print
        result = highwayConstruction(n, k)
        print result

